### DISCUSSION TITLE
Is there an API that can get 29, 30 and 31 diagnostic TLVs data through a single request?

### DISCUSSION THREAD

**Discussion by jinpeng1989** [openthread/openthread#9913] [Q&A] [answered]
I expect to get 29, 30 and 31 diagnostic TLVs data through a single request.
I did some experiments where I could get this data separately though meshdiag.
```
root@KNBA5F000300:~# ot-ctl meshdiag childtable 0x4000
rloc16:0x4002 ext-addr:ea6c419803df76bb ver:2
    timeout:3300 age:124 supvn:129 q-msg:1
    rx-on:no type:mtd full-net:yes
    rss - ave:-26 last:-27 margin:94
    err-rate - frame:0.63% msg:0.00% 
    conn-time:01:39:10
    csl - sync:no period:0 timeout:0 channel:0
Done
root@KNBA5F000300:~# ot-ctl meshdiag childip6 0x4000
child-rloc16: 0x4002
    fdfd:b182:4277:1297:fc4f:36b8:9f79:ff9
    fd0c:179e:644b:1:1f37:faa:c5b7:19a4
Done
root@KNBA5F000300:~# ot-ctl meshdiag routerneighbortable 0x4000
rloc16:0xe000 ext-addr:22cbf7eace20d1a3 ver:4
    rss - ave:-42 last:-42 margin:78
    err-rate - frame:1.28% msg:1.53% 
    conn-time:01:48:39
Done
```
![image](https://github.com/openthread/openthread/assets/97089879/17d5affd-31d2-4f88-99ef-87d040494663)

I tried to use the networkdiagnostic to get  29, 30 and 31 diagnostic TLVs data and the Thread Router responded with no data.
```
root@KNBA5F000300:~# ot-ctl networkdiagnostic get fdfd:b182:4277:1297:0:ff:fe00:4000 29 30 31
DIAG_GET.rsp/ans from fdfd:b182:4277:1297:0:ff:fe00:4000: 
Done
```
![image](https://github.com/openthread/openthread/assets/97089879/e5294ef6-fc3f-47f0-90b6-5c185fcf375a)

And then I tried to use the networkdiagnostic to get  0, 1 and 6 diagnostic TLVs data and the Thread Router responded with data successfully. It looks like the stack is discriminating against these new TLVs. Why?
```
root@KNBA5F000300:~# ot-ctl networkdiagnostic get fdfd:b182:4277:1297:0:ff:fe00:4000 0 1 6
DIAG_GET.rsp/ans from fdfd:b182:4277:1297:0:ff:fe00:4000: 0008567aacc880b7b4810102400006084cebf5e940a89810
Ext Address: '567aacc880b7b481
Rloc16: 0x4000
Leader Data:
    PartitionId: 0x4cebf5e9
    Weighting: 64
    DataVersion: 168
    StableDataVersion: 152
    LeaderRouterId: 0x10
Done
```
![image](https://github.com/openthread/openthread/assets/97089879/03c1067a-ef24-40df-8e0d-7a76482ca66f)

**Comment by abtink** [comment]
> I expect to get 29, 30 and 31 diagnostic TLVs data through a single request.
> I did some experiments where I could get this data separately though meshdiag.
> It looks like the stack is discriminating against these new TLVs. Why?

You can use the `meshdiag` commands. Please see [`mesh_diag.h`](https://github.com/openthread/openthread/blob/main/include/openthread/mesh_diag.h)

The new network diag TLVs function differently. They're list-based, meaning the answer can potentially contain numerous entries (e.g., a device might have up to 512 children or multiple addresses) which may not fit in a single IPv6 message MTU.
- To accommodate this, the design defines a mechanism for sending responses across multiple TMF Diag Answer messages. Additional TLVs indicate the relevant query and whether more information is forthcoming.
- The `MeshDiag` module handles this complexity, providing collected information directly to the user. This simplifies the process, removing the need for users of API to track and associate multiple answers with individual queries.

BWT, this question is similar to https://github.com/orgs/openthread/discussions/9302.

**Reply by jinpeng1989** [reply]
Thank you, I understand the design, defines a mechanism for sending responses across multiple TMF Diag Answer messages. 
This mechanism is also capable of responding to multiple TLVs. So, I recommend adding Get Diagnostic API Query API that support request multiple specified TLVs.

**Comment by jinpeng1989** [comment]
Why not use the CoAP Block2 Option? It can also solve the problem of very long packets.

**Reply by abtink** [reply, marked as answer]
Many ideas including CoAP block-wise were considered and discussed in Thread Technical group during the design of the new mechnasim.
- You can read more details in [SPEC-1151](https://threadgroup.atlassian.net/browse/SPEC-1151) and other JIRAs.
- All the related PRs and SPEC JIRAs are listed in https://github.com/openthread/openthread/issues/8460

Quick answer on CoAP block-wise:
- Code size overhead: Both sender and receiver would need to support CoAP block options (at least Block2). Adding this functionality will increase code size.
- Block boundaries don't necessarily align with TLVs, complicating the preparation and parsing process. Receivers may need to collect multiple blocks before parsing, impacting memory efficiency.
- The block-wise CoAP support in OT was an external contribution and may lack the extensive testing of other CoAP functions. I wouldn't personally recommend its use in a production environment without careful consideration.

**Comment by jinpeng1989** [comment]
Currently, our product takes four get diagnostic request/query for a Thread Router. There are 15 Thread Routers in system, which means that 60 diagnostic requests are required to obtain complete diagnostic data. Is there any way to reduce the number of packets exchanged?
![image](https://github.com/user-attachments/assets/9383f6d2-4485-4ffc-ac3d-721bed835fb1)

**Comment by jinpeng1989** [comment]
Is it possible to extend the mesh dianostic module so that it can get 29, 30, 31 TLVs in one Query/Answer?

**Reply by jwhui** [reply]
It is technically possible, but would lead to really large response messages. It's not clear to me whether combining them all into one message will improve overall performance. If you are interested in pursuing this, I would suggest comparing the performance before and after.

**Comment by venkatadivvi** [comment]
Hi,
I am new to this group, I am searching for any existing discussion on newly added Thread 1.4 TLV's ( Answer/Query 32 & 33), looking for any sample CLI outputs or packet traces for cases where responses for "network diagnostics" provided in multiple sequential "answer(s)". 
Is there any timer for completing the reception of all answers (e.g. For a requested query, node sending 15 answers, with L0/Index-1,Query-ID aabb to L1/Index-15,Query-ID aabb). 
Could you please redirect me to any existing discussions/resources referring the implementation examples.
Thank you.
Regards,
Ramana.

**Reply by abtink** [reply]
@venkatadivvi, this is all handled by the OpenThread `MeshDiag` module. The idea is that users don't need to deal with these TLVs directly, making it much easier for them to retrieve the information they want using simple `otMeshDiag` APIs.

* You can check the [`otMeshDiag` APIs](https://github.com/openthread/openthread/blob/main/include/openthread/mesh_diag.h) and its related [CLI `meshdiag` commands](https://github.com/openthread/openthread/blob/f600294c436cdae6044cc6d67d1a9dfd6ca44ee5/src/cli/README.md#meshdiag-topology-ip6-addrs-children))

* If you are interested in the specific way these TLVs are used, you can check the implementation of the `MeshDiag` module ([`mesh_diag.hpp`](https://github.com/openthread/openthread/blob/main/src/core/utils/mesh_diag.hpp) and [`mesh_diag.cpp`](https://github.com/openthread/openthread/blob/main/src/core/utils/mesh_diag.cpp)).

* Here are some other similar questions/discussions on this:
    * https://github.com/orgs/openthread/discussions/9302#discussioncomment-6490745
    * https://github.com/orgs/openthread/discussions/9913#discussioncomment-8749719

**Comment by venkatadivvi** [comment]
Thanks, Abtink.
Yes, its good command/API to get the topology but its not giving the complete list of other key diagnostic TLV's, which are important for operations aspect. Operations might be interested in specific set of TLV's , in TLV format to integrate with upstream systems designed for network monitoring aspect.
Trying to see how these commands gets processed and responded:
networkdiagnostic get <ipv6_address> 6 7 8 9 14 15 16
networkdiagnostic get <ipv6_address> 29

6: Leader Data TLV
7: Network Data TLV
8: IPv6 Address List TLV
9: MAC Counters TLV
14: Battery Level TLV
15: Supply Voltage TLV
16: Child Table TLV
29: Child TLV

In a simple one parent and two child thread network, the above command requesting three TLV's. And response payload size would be bigger than 127 bytes and  I guess it would be responded as multi-Answer TLV method. Would like to see how those packets looks like and processed.

**Reply by abtink** [reply]
You can use `otThreadSendDiagnosticGet()` for other TLVs (which contain single-value or short list).

It seems like you want to learn the details. I would suggest studying the Thread specification, which goes over the details on these TLVs and all commands. I can give you a short overview to get you started:

* Note that there are different TMF commands, and there are restrictions on which TLVs can be used with which command:
    * `Network Diagnostic Get Request ("d/dg")`, which results in an immediate CoAP response and can be used with most of the earlier TLVs (the TLVs that have a single value or short list).
    * `Network Diagnostic Get Query ("d/dq")`, which can be used to query multiple devices at once (multicast) or used to query new list-based TLVs (where the result may contain many items). The `GetQuery` command response itself uses a separate TMF exchange, `Network Diagnostic Get Answer ("d/da")`, and can result in one or more "Answer" messages (thus allowing us to retrieve larger diagnostic information from a device such as detailed child or neighbor tables).
    * All the new list-based TLVs which can contain a larger number of items can only be used with `Get Query` and `Get Answer` TMF messages: These include "Child TLV (29)", "ChildIp6AddressList TLV (30)", "RouterNeighbor (31)", "Answer (32)", and "QueryID (33)".

So, `networkdiagnostic get <ipv6_address> 29` (with a unicast address) is not a valid command, as it would use `Network Diagnostic Get Request` TMF command, which is not valid for the Child TLV (29).

The `MeshDiag` API can be used for these TLVs (e.g., `otMeshDiagQueryChildTable()`). They handle sending the correct `GetQuery` TMF command and processing all the associated `Get Answer` responses for this query, giving the user back all the parsed information. As I said, the idea is to make it easier for the user.

**Reply by abtink** [reply]
Feel free to read the code in the following modules if you want to learn how these messages are generated and processed:
- https://github.com/openthread/openthread/blob/main/src/core/thread/network_diagnostic.hpp
- https://github.com/openthread/openthread/blob/main/src/core/thread/network_diagnostic.cpp
- https://github.com/openthread/openthread/blob/main/src/core/utils/mesh_diag.hpp and
- https://github.com/openthread/openthread/blob/main/src/core/utils/mesh_diag.cpp
-

**Comment by venkatadivvi** [comment]
Thank you, real useful.
